### PR TITLE
DM-8465: wrap meas_modelfit with pybind11

### PR DESCRIPTION
**Pull request by TallJimbo**

### CODE REVIEW

File: python/lsst/meas/modelfit/model.cc [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        "center"_a, "radius"_a=1.0
+    );
+    cls.def("getNonlinearDim", &Model::getNonlinearDim);
+    cls.def("getAmplitudeDim", &Model::getAmplitudeDim);
+    cls.def("getFixedDim", &Model::getFixedDim);
+    cls.def("getBasisCount", &Model::getBasisCount);
+    cls.def("getNonlinearNames", &Model::getNonlinearNames);
+    cls.def("getAmplitudeNames", &Model::getAmplitudeNames);
+    cls.def("getFixedNames", &Model::getFixedNames);
+    cls.def("getBasisVector", &Model::getBasisVector);
```
Comment (pschella) [inline python/lsst/meas/modelfit/model.cc]: The above four functions return by `const &`. Perhaps make them return by value instead?

File: python/lsst/meas/modelfit/mixture.cc [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include "pybind11/pybind11.h"
+#include "pybind11/stl.h"
+
+#include <sstream>  // Python.h must come before even system headers
```
Comment (pschella) [inline python/lsst/meas/modelfit/mixture.cc]: I don't think we do this consistently.

File: python/lsst/meas/modelfit/mixture.cc [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+namespace lsst {
+namespace meas {
+namespace modelfit {
+namespace {
+
+using PyMixtureComponent = py::class_<MixtureComponent>;
```
Comment (pschella) [inline python/lsst/meas/modelfit/mixture.cc]: `unique_ptr` on purpose?

File: python/lsst/meas/modelfit/mixture.cc [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    cls.def(py::init<int>(), "dim"_a);
+    cls.def(py::init<Scalar, Vector const &, Matrix const &>(), "weight"_a, "mu"_a, "sigma"_a);
+    cls.def(
+        "__str__",
+        [](MixtureComponent const & self) {
+            std::ostringstream os;
+            os << self;
+            return os.str();
+        }
+    );
```
Comment (pschella) [inline python/lsst/meas/modelfit/mixture.cc]: Perhaps we should make a utility function for this (and `__repr__`).

File: python/lsst/meas/modelfit/mixture.cc [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+namespace modelfit {
+namespace {
+
+using PyMixtureComponent = py::class_<MixtureComponent>;
+using PyMixtureUpdateRestriction = py::class_<MixtureUpdateRestriction,
+                                              std::shared_ptr<MixtureUpdateRestriction>>;
+using PyMixture = py::class_<Mixture, std::shared_ptr<Mixture>,
+                             afw::table::io::PersistableFacade<Mixture>, afw::table::io::Persistable>;
+
+PyMixtureComponent declareMixtureComponent(py::module & mod) {
```
Comment (pschella) [inline python/lsst/meas/modelfit/mixture.cc]: We seem to have decided not to make these static. I shall update the documentation.

File: python/lsst/meas/modelfit/mixture.cc [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+            ndarray::Array<Scalar,1,0> const & array
+        ) -> Scalar {
+            return self.evaluate(component, array.asEigen());
+        },
+        "component"_a, "x"_a
+    );
+    cls.def(
+        "evaluate",
+        [](Mixture const & self, ndarray::Array<Scalar,1,0> const & array) -> Scalar {
+            return self.evaluate(array.asEigen());
```
Comment (pschella) [inline python/lsst/meas/modelfit/mixture.cc]: Souldn't Eigen input just work too?

File: python/lsst/meas/modelfit/mixture.cc (hunk 5 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/mixture.cc]: No, I was just uniformly forgetful on this ticket.  I'll go and make them all static.

File: python/lsst/meas/modelfit/priors/priors.cc [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+    if (_import_array() < 0) {
+        PyErr_SetString(PyExc_ImportError, "numpy.core.multiarray failed to import");
+        return nullptr;
+    }
+
+    declarePrior(mod);
+    declareMixture(mod);
+    declareSemiEmpirical(mod);
+    declareSoftenedLinear(mod);
```
Comment (pschella) [inline python/lsst/meas/modelfit/priors/priors.cc]: I would suffix these with `Prior` to match the class names, but not strictly required by our rules.
Also, while I think mixing these in one wrapper file is not actually needed here it also doesn't seem so bad. But others may have a different opinion.

File: tests/testMixture.py [hunk 8]
```diff
@@ -70,7 +70,6 @@ def testSwig(self):
         l1[2].weight = 0.5
         m1 = lsst.meas.modelfit.Mixture(1, l1)
         self.assertEqual(m1[0].weight, 0.5)
-        self.assertEqual(len(l1), 0)
```
Comment (pschella) [inline tests/testMixture.py]: on purpose?

Comment (pschella) [inline tests/testMixture.py]: I assume so, given that l1 is now a Python list.

File: tests/testProjectedLikelihood.py [hunk 9]
```diff
@@ -191,8 +192,7 @@ def testDirect(self):
         ctrl = lsst.meas.modelfit.UnitTransformedLikelihoodControl()
         var = numpy.random.rand(self.bbox0.getHeight(), self.bbox0.getWidth()) + 2.0
         self.exposure0.getMaskedImage().getVariance().getArray()[:, :] = var
-        efv = lsst.meas.modelfit.EpochFootprintVector()
-        efv.push_back(lsst.meas.modelfit.EpochFootprint(self.footprint0, self.exposure0, self.psf0))
+        efv = [lsst.meas.modelfit.EpochFootprint(self.footprint0, self.exposure0, self.psf0)]
```
Comment (pschella) [inline tests/testProjectedLikelihood.py]: I'm really liking actually having Python list and tuple types. It looks and acts so much better. Although I'm sure there is a performance penalty somewhere which we have yet to identify and workaround with `ndarray`.

File: python/lsst/meas/modelfit/truncatedGaussian.cc [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+using Evaluator = TruncatedGaussianEvaluator;
+using LogEvaluator = TruncatedGaussianLogEvaluator;
+
+using PyTruncatedGaussian = py::class_<TruncatedGaussian, std::shared_ptr<TruncatedGaussian>>;
+using PySampler = py::class_<Sampler, std::shared_ptr<Sampler>>;
+using PyEvaluator = py::class_<Evaluator, std::shared_ptr<Evaluator>>;
+using PyLogEvaluator = py::class_<LogEvaluator, std::shared_ptr<LogEvaluator>>;
+
+template <typename Class, typename PyClass>
+PyClass declareEvaluator(py::module & mod, std::string const & name) {
```
Comment (r-owen) [inline python/lsst/meas/modelfit/truncatedGaussian.cc]: The code being wrapped is surprising in that wraps two unrelated classes that apparently happen to have the same API. Thus some documentation would be really helpful, e.g. explaining that it wraps two different classe: TruncatedGaussianEvaluator and TruncatedGaussianLogEvaluator, and that:
- `name` is one of "Evaluator" or "LogEvaluator".
- This does not wrap `template <typename Derived> Scalar operator()` because...(why not? I really don't know)

File: python/lsst/meas/modelfit/truncatedGaussian.cc [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+PYBIND11_PLUGIN(truncatedGaussian) {
+
+    py::module mod("truncatedGaussian");
+
+    if (_import_array() < 0) {
+        PyErr_SetString(PyExc_ImportError, "numpy.core.multiarray failed to import");
+        return nullptr;
+    }
+
+    py::module::import("lsst.afw.math");
```
Comment (r-owen) [inline python/lsst/meas/modelfit/truncatedGaussian.cc]: We usually put import statements appeare before `py::module mod(...` as per the example in https://developer.lsst.io/v/DM-8577/coding/pybind11_style_guide.html#id11

Does it make any practical difference where it goes, e.g. does it give more informative and less confusing error messages in one location vs. the other?

All else being equal it'd be nice if we could pick a spot and stick to it, and not have that spot vary depending on whether we test for numpy.core.

File: python/lsst/meas/modelfit/truncatedGaussian.cc [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    cls.def("evaluateLog", &TruncatedGaussian::evaluateLog);
+    cls.def("evaluate", &TruncatedGaussian::evaluate);
+    cls.def("getDim", &TruncatedGaussian::getDim);
+    cls.def("maximize", &TruncatedGaussian::maximize);
+    cls.def("getUntruncatedFraction", &TruncatedGaussian::getUntruncatedFraction);
+    cls.def("getLogPeakAmplitude", &TruncatedGaussian::getLogPeakAmplitude);
+    cls.def("getLogIntegral", &TruncatedGaussian::getLogIntegral);
+
+    cls.attr("LogEvaluator") = declareEvaluator<LogEvaluator, PyLogEvaluator>(mod, "LogEvaluator");
+    cls.attr("Evaluator") = declareEvaluator<Evaluator, PyEvaluator>(mod, "Evaluator");
```
Comment (r-owen) [inline python/lsst/meas/modelfit/truncatedGaussian.cc]: Why not `cls.attr("Sampler") =...`? Please add a code comment explaining why this is missing, or, perhaps, why the above two are present.

File: python/lsst/meas/modelfit/sampler.cc [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include "pybind11/pybind11.h"
+
```
Comment (r-owen) [inline python/lsst/meas/modelfit/sampler.cc]: (Delete blank line)

File: tests/testCModel.py [hunk 14]
```diff
@@ -43,7 +43,7 @@ def makeMultiShapeletCircularGaussian(sigma):
     s = lsst.shapelet.ShapeletFunction(0, lsst.shapelet.HERMITE, sigma)
     s.getCoefficients()[0] = 1.0 / lsst.shapelet.ShapeletFunction.FLUX_FACTOR
     m = lsst.shapelet.MultiShapeletFunction()
-    m.getComponents().push_back(s)
+    m.addComponent(s)
```
Comment (kfindeisen) [inline tests/testCModel.py]: It's  obviously better-encapsulated this way, but how is this pybind11 related? I don't see an `addComponent` method elsewhere in the diff.

File: python/lsst/meas/modelfit/integrals.cc [hunk 15]
```diff
@@ -21,21 +21,23 @@
...
+namespace meas {
+namespace modelfit {
+namespace {
 
-%pythoncode %{
-import lsst.pex.config
+PYBIND11_PLUGIN(integrals) {
+    py::module mod("integrals");
+    mod.def("phid", &detail::phid);
+    mod.def("bvnu", &detail::bvnu);
```
Comment (r-owen) [inline python/lsst/meas/modelfit/integrals.cc]: This seems to define `lsst.meas.multifit.integrals.bvnu` yet when you reference it in tests/testIntegrals.py you reference it as `lsst.meas.multifit.integrals.detail.bvnu`. Could you please explain what's going on?

File: tests/testGeneralShapeletPsfApproxPlugins.py [hunk 16]
```diff
@@ -34,6 +34,7 @@
 import lsst.log.utils
 import lsst.meas.modelfit
 import lsst.meas.base
+import lsst.meas.algorithms
```
Comment (kfindeisen) [inline tests/testGeneralShapeletPsfApproxPlugins.py]: As above.

File: tests/testDoubleShapeletPsfApprox.py [hunk 17]
```diff
@@ -34,6 +34,7 @@
 import lsst.log
 import lsst.log.utils
 import lsst.meas.modelfit
+import lsst.meas.algorithms
```
Comment (kfindeisen) [inline tests/testDoubleShapeletPsfApprox.py]: Why wasn't this import needed before?

File: tests/testPsfFitter.py [hunk 18]
```diff
@@ -123,7 +123,9 @@ def testFixedModel(self):
         ellipses3 = model.makeEllipseVector()
         for i in range(len(ellipses2)):
             self.assertFloatsAlmostEqual(ellipses1[i].getParameterVector(), ellipses2[i].getParameterVector())
-            ellipses3[i] = msf.getComponents()[i].getEllipse()  # need to convert ellipse parametrization
+            # need to convert ellipse parametrization
+            ellipses3[i].setCore(msf.getComponents()[i].getEllipse().getCore())
+            ellipses3[i].setCenter(msf.getComponents()[i].getEllipse().getCenter())
```
Comment (kfindeisen) [inline tests/testPsfFitter.py]: Why doesn't the old code work? Is it because python doesn't have C++-like assignment? Again, I don't see any changes in the wrapping that are related.

Same question for the two analogous blocks later in the file.

File: python/lsst/meas/modelfit/mixture.cc (hunk 4 above)
Comment (r-owen) [inline python/lsst/meas/modelfit/mixture.cc]: How about a function `template <class T> toString(T const &) {...}` that uses `<<`? Then it could be freely used for "__str__" and "__repr__". This would fit neatly in `utils/pybind11.h` (though that name needs updating to meet our standards).

File: python/lsst/meas/modelfit/mixture.cc (hunk 2 above)
Comment (r-owen) [inline python/lsst/meas/modelfit/mixture.cc]: Agreed. I have never done this. Are you sure it is necessary?

File: python/lsst/meas/modelfit/mixture.cc (hunk 4 above)
Comment (kfindeisen) [inline python/lsst/meas/modelfit/mixture.cc]: Would be nice, but we don't have a convention yet for the relationships between `__str__`, `__repr__`, and `operator<<`. This was brought up on [CLO](https://community.lsst.org/t/methods-for-stringification/1578/5) but didn't get anywhere.

File: python/lsst/meas/modelfit/model.cc (hunk 1 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/model.cc]: They all return `std::vector`->`list`, so I don't think it matters to the compiler.  It may matter to humans; do you think we should use `return_value_policy::copy` to make it explicit that that's happening here?

File: python/lsst/meas/modelfit/mixture.cc (hunk 2 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/mixture.cc]: Python docs say it is necessary: https://docs.python.org/2/c-api/intro.html#include-files

The problems may only appear on operating systems or hardware we don't support, but it's definitely something we ought to be doing.

File: python/lsst/meas/modelfit/mixture.cc (hunk 3 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/mixture.cc]: Yes.  Small helper class, virtually no chance of being used in wrappers other than those defined here.

File: python/lsst/meas/modelfit/mixture.cc (hunk 4 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/mixture.cc]: @kfindeisen's point about us not having standardized what goes where is why I haven't been motivated to add a generic function for this yet.

File: python/lsst/meas/modelfit/mixture.cc (hunk 6 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/mixture.cc]: Yes, but that might require a lambda too (since the original C++ signature takes a CRTP base class), and it would require an extra copy.

File: tests/testMixture.py (hunk 8 above)
Comment (TallJimbo) [inline tests/testMixture.py]: Right.  This used to consume the list's elements in Python, in what was almost certainly a premature optimization that tried to avoid copying.

File: tests/testProjectedLikelihood.py (hunk 9 above)
Comment (TallJimbo) [inline tests/testProjectedLikelihood.py]: I'm worried about the performance penalties we *can't* easily work around with ndarray (because the item type isn't a numpy type).  I haven't seen any cases that worry me.

And in fairness to Swig, I probably could have used the same syntax for these lines back in the Swig days, and if I wasn't it was because I defined the vector wrappers and the wrappers for this function in the wrong order.  But I do agree with you in general.

File: python/lsst/meas/modelfit/adaptiveImportanceSampler.cc [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+    LSST_DECLARE_CONTROL_FIELD(clsImportanceSamplerControl, ImportanceSamplerControl, targetPerplexity);
+    LSST_DECLARE_CONTROL_FIELD(clsImportanceSamplerControl, ImportanceSamplerControl, maxRepeat);
+
+    PyAdaptiveImportanceSampler clsAdaptiveImportanceSampler(mod, "AdaptiveImportanceSampler");
+    clsAdaptiveImportanceSampler.def(
+        py::init<afw::table::Schema &, std::shared_ptr<afw::math::Random>,
+                 std::map<int, ImportanceSamplerControl> const &, bool>(),
+        "sampleSchema"_a, "rng"_a, "ctrls"_a, "doSaveIteration"_a=false
+    );
+    // virtual run method already wrapped by Sampler base class
```
Comment (TallJimbo) [inline python/lsst/meas/modelfit/adaptiveImportanceSampler.cc]: I was on the fence about adding `override` on this ticket, but you wanting it too settles it.

I don't have any concern about these working.  I don't believe we've ever had a case of inherited virtual methods not working - the case I suspect you're thinking of (`BaseRecord::getTable`, right?) was non-virtual methods that delegate to virtual methods and cast the types.  And there the automatic type-casting only wasn't working (I'm pretty certain) because the *actual* type returned wasn't either the base class or the exposed derived type (it was an even-further-derived type).

File: python/lsst/meas/modelfit/optimizer/optimizer.cc [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+using PyOptimizerHistoryRecorder =
+    py::class_<OptimizerHistoryRecorder, std::shared_ptr<OptimizerHistoryRecorder>>;
+using PyOptimizer = py::class_<Optimizer, std::shared_ptr<Optimizer>>;
+
+PyOptimizerObjective declareOptimizerObjective(py::module & mod) {
+    PyOptimizerObjective cls(mod, "OptimizerObjective");
+    cls.def_readonly("dataSize", &OptimizerObjective::dataSize);
+    cls.def_readonly("parameterSize", &OptimizerObjective::parameterSize);
+    cls.def_static("makeFromLikelihood", &OptimizerObjective::makeFromLikelihood,
+                   "likelihood"_a, "prior"_a=py::none());
```
Comment (TallJimbo) [inline python/lsst/meas/modelfit/optimizer/optimizer.cc]: And now that I think about it, we should probably use `nullptr` instead of either of those.

I'll add the comment.

File: python/lsst/meas/modelfit/optimizer/optimizer.cc [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+    cls.def_readonly("prior", &OptimizerHistoryRecorder::prior);
+    cls.def_readonly("trust", &OptimizerHistoryRecorder::trust);
+    cls.def_readonly("parameters", &OptimizerHistoryRecorder::parameters);
+    cls.def_readonly("derivatives", &OptimizerHistoryRecorder::derivatives);
+    return cls;
+}
+
+PyOptimizer declareOptimizer(py::module & mod) {
+    PyOptimizer cls(mod, "Optimizer");
+    // StateFlags enum is used as bitflag, so we wrap values as int class attributes.
```
Comment (TallJimbo) [inline python/lsst/meas/modelfit/optimizer/optimizer.cc]: I actually think we need to switch this convention to wrap them as integer attributes, and I had been planning to make that change in most other places on the cleanup tickets.  I don't think there's any value to making bitflag enums appear as enums in Python.

File: python/lsst/meas/modelfit/optimizer/optimizer.cc [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+                 Optimizer::Control>(),
+        "objective"_a, "parameters"_a, "ctrl"_a
+    );
+    cls.def("getObjective", &Optimizer::getObjective);
+    cls.def("getControl", &Optimizer::getControl, py::return_value_policy::copy);
+    cls.def(
+        "step",
+        (bool (Optimizer::*)(Optimizer::HistoryRecorder const &, afw::table::BaseCatalog &)) &Optimizer::step,
+        "recorder"_a, "history"_a);
+    cls.def("step", (bool (Optimizer::*)()) &Optimizer::step);
```
Comment (TallJimbo) [inline python/lsst/meas/modelfit/optimizer/optimizer.cc]: Accident.  I'll put them back.

File: python/lsst/meas/modelfit/truncatedGaussian.cc [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include "pybind11/pybind11.h"
+#include "pybind11/stl.h"
+
```
Comment (TallJimbo) [inline python/lsst/meas/modelfit/truncatedGaussian.cc]: Is that actually required? I think it's actually more readable to group headers from a single package (on a case-by-case basis, at least).

File: python/lsst/meas/modelfit/truncatedGaussian.cc (hunk 11 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/truncatedGaussian.cc]: It doesn't matter.  I'll move them all before the `py::module` declaration.

File: python/lsst/meas/modelfit/truncatedGaussian.cc (hunk 10 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/truncatedGaussian.cc]: Comments added. The other overload of `operator()` was not wrapped because it's just an Eigen version of the ndarray one, and hence they look the same in Python.

File: python/lsst/meas/modelfit/truncatedGaussian.cc (hunk 12 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/truncatedGaussian.cc]: It is present, a few lines later (after the `Sampler` wrappers are defined).

File: python/lsst/meas/modelfit/integrals.cc (hunk 15 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/integrals.cc]: It's actually referenced as `lsst.meas.modelfit.detail.bvnu`, which works because `meas/modelfit/__init__.py` has:
```
from . import integrals as detail
```
I'm not sure that was wise, but it nicely makes the Python namespace match the C++ one while letting the wrapper source file match the header name.

File: include/lsst/meas/modelfit/DoubleShapeletPsfApprox.h [hunk 24]
```diff
@@ -70,7 +70,7 @@ class DoubleShapeletPsfApproxControl {
     );
 
     LSST_NESTED_CONTROL_FIELD(
-        optimizer, lsst.meas.modelfit.modelfitLib, OptimizerControl,
+        optimizer, lsst.meas.modelfit, OptimizerControl,
```
Comment (TallJimbo) [inline include/lsst/meas/modelfit/DoubleShapeletPsfApprox.h]: I think that should work, and it would indeed be more consistent with what I've done elsewhere.  I'll give it a try.

I'm afraid all of this control field wrapping is a bit of a fragile mess right now; see comments on [DM-9222](https://jira.lsstcorp.org/browse/DM-9222).  It's not a regression from Swig (it was also a mess before), but I hope pybind11 will give us an opportunity to fix it.

File: python/lsst/meas/modelfit/__init__.py [hunk 25]
```diff
@@ -20,7 +20,21 @@
 # see <http://www.lsstcorp.org/LegalNotices/>.
 #
 from .version import *
-from .modelfitLib import *
+from .common import *
+from .mixture import *
+from .unitSystem import *
 from .priors import *
-from . import psf     # import these into separate namespaces because they just contain plugins,
-from . import cmodel  # which we don't expect users to need to get to outside of their registries.
```
Comment (TallJimbo) [inline python/lsst/meas/modelfit/__init__.py]: There are many symbols defined in `cmodel` and `psf` that were previously defined in `modelfitLib` and hence lifted to `lsst.meas.modelfit`, and those need to be public.  Putting them in a subpackage with the relatively small amount of pure-Python `psf` and `cmodel` code greatly improved the overall organization of this package, and it made it seem silly to treat that pure-Python code differently here.

File: python/lsst/meas/modelfit/cmodel/cmodel.cc [hunk 26]
```diff
@@ -0,0 +1,235 @@
...
+    LSST_DECLARE_NESTED_CONTROL_FIELD(cls, CModelControl, exp);
+    LSST_DECLARE_NESTED_CONTROL_FIELD(cls, CModelControl, dev);
+    LSST_DECLARE_CONTROL_FIELD(cls, CModelControl, minInitialRadius);
+    LSST_DECLARE_CONTROL_FIELD(cls, CModelControl, fallbackInitialMomentsPsfFactor);
+    return cls;
+}
+
+// Custom wrapper for views to std::bitset.
+template <int N>
+class BitSetView {
```
Comment (TallJimbo) [inline python/lsst/meas/modelfit/cmodel/cmodel.cc]: pybind11 also has lacks support for `std::bitset`. so wrapping the `flags` attributes directly wasn't an option.  I could have proceeded with the same workaround I had for Swig (the `getFlag`/`setFlag` methods), but I preferred the `.flags[...]` approach, and now that it was relatively easy to do that in Python (though not trivial), I decided to delete the old one.  That was before I realized how much code (including C++ code) was using what I considered the Swig-workaround interface, but I still think it's an improvement to have only one way to get and set these flags and to have it look similar across languages.

File: python/lsst/meas/modelfit/mixture.cc (hunk 4 above)
Comment (r-owen) [inline python/lsst/meas/modelfit/mixture.cc]: A generic function could be used for str or repr; it seems highly likely at least one of those will be able to use `ostringstream <<` (it seems a bit mad to do anything else).

File: python/lsst/meas/modelfit/optimizer/optimizer.cc (hunk 21 above)
Comment (r-owen) [inline python/lsst/meas/modelfit/optimizer/optimizer.cc]: You are saying that in general if an enum is used to define a bitmask that the bitmask type should be int-like instead of the enum class? Interesting. I like the type safety of having to use named constants, but it is, admittedly, a headache when specifying a value in a config (as for all enums, due to problems with the config design) and I it is difficult to check when persisting and unpersisting. Perhaps this is a case where pybind11 is unhelpful in its type safety; being able to define such objects using enum seems natural but makes life hard for us.

File: python/lsst/meas/modelfit/truncatedGaussian.cc (hunk 23 above)
Comment (r-owen) [inline python/lsst/meas/modelfit/truncatedGaussian.cc]: It doesn't seem to be required (the grouping rules are much vaguer than I remembered). In any case readability counts. Leave it alone. As an aside, I don't think we are obeying the order rules (nor am I convinced they are reasonable) https://developer.lsst.io/coding/cpp_style_guide.html#include-statements-should-be-sorted-and-grouped

File: python/lsst/meas/modelfit/optimizer/optimizer.cc (hunk 21 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/optimizer/optimizer.cc]: I think of using `enum` to define a bitflag in C++ as at best a form of shorthand for defining the same symbols as integer constants, and at worst (in the era of C++11 with `constexpr`) a bit of a hack we should consider moving away from.  Note that with bitflags you don't even get any type safety unless you enumerate all possible combination of bits (which is a bad idea), since you always have to pass the values via some sort of integer.

File: python/lsst/meas/modelfit/integrals.cc (hunk 15 above)
Comment (r-owen) [inline python/lsst/meas/modelfit/integrals.cc]: Ouch. I had not even noticed that the entire contents of `integrals.h` is defined in sub-namespace `detail`. I am very uncomfortable with that cheat, especially since the `detail` directory exists and `integrals.h` could easily be moved there.

If you are willing to clean that up I suggest moving the C++ header file (it is only included by 4 files, all of them in meas_multifit) and making a `detail` subdirectory in the python dir.

File: python/lsst/meas/modelfit/cmodel/cmodelContinued.py [hunk 27]
```diff
@@ -21,18 +20,24 @@
...
 #
 
-import lsst.pex.config
+from .cmodel import CModelStageControl, CModelControl, CModelAlgorithm
+
+from lsst.pex.config import makeConfigClass
 import lsst.meas.base
-from . import modelfitLib
 
+__all__ = ("CModelStageConfig", "CModelConfig")
```
Comment (TallJimbo) [inline python/lsst/meas/modelfit/cmodel/cmodelContinued.py]: The code in this particular file is a happy exception to the "mostly untested" statement.  The plugin classes are added to a registry in `lsst.meas.base` (via the `register` decorator), and are used exclusively from that registry, so there's no need to import them directly from this module.  I see that I added a comment to that effect in `psfContinued.py` but not here; I'll add one here, too.

Comment (kfindeisen) [inline python/lsst/meas/modelfit/cmodel/cmodelContinued.py]: Great! :grinning:

File: python/lsst/meas/modelfit/psf/psfContinued.py [hunk 28]
```diff
@@ -24,46 +23,67 @@
...
+        itemtype=GeneralPsfFitterConfig,
         doc="a dictionary of models that can be used to fit the PSF",
-        default={} # populated in setDefaults; can't do it on a single line here
+        default={} # populated in setDefaults; can't do it on a single line
     )
     sequence = lsst.pex.config.ListField(
         dtype=str,
-        doc="a sequence of model names indicating which models should be fit, and their order",
+        doc=("a sequence of model names indicating which models should be fit,"
+             " and their order"),
```
Comment (TallJimbo) [inline python/lsst/meas/modelfit/psf/psfContinued.py]: 79-character line limit.

File: python/lsst/meas/modelfit/psf/psf.cc [hunk 29]
```diff
@@ -0,0 +1,222 @@
...
+            afw::image::Image<Pixel> const & image,
+            afw::geom::ellipses::Quadrupole const & moments,
+            Scalar noiseSigma
+        ) {
+            return self.apply(image, moments, noiseSigma);
+        },
+        "image"_a,
+        "moments"_a,
+        "noiseSigma"_a=-1
+    );
```
Comment (TallJimbo) [inline python/lsst/meas/modelfit/psf/psf.cc]: The original signature has an optional argument (`int*` used as an optional output) not relevant for Python.  Comment added to that effect.

File: tests/testCModel.py (hunk 14 above)
Comment (TallJimbo) [inline tests/testCModel.py]: This is adapting to a necessary change made by @pschella when wrapping the `shapelet` package.  This object previously exposed a non-const reference to a STL container that needed to be modified in place.  That's no longer possible now that STL containers are only wrapped as copies.

File: tests/testGeneralShapeletPsfApproxPlugins.py (hunk 16 above)
Comment (TallJimbo) [inline tests/testGeneralShapeletPsfApproxPlugins.py]: `lsst.meas.modelfit` previously imported code that imported `lsst.meas.algorithms`, and since that just puts the `algorithms` subpackage into the existing `lsst.meas` package due to our unusual import hooks, the need for `lsst.meas.algorithms` here was disguised.

File: tests/testPsfFitter.py (hunk 18 above)
Comment (TallJimbo) [inline tests/testPsfFitter.py]: You guessed it.  The old code relied on invoking a C++ assignment operator that converted from the source ellipse parameterization to the destination one.  That used to work okay through the Swig-wrapped `std::vector`s, but it doesn't work when assigning to a `list`.

Clearly, this pattern is quite fragile in Python, and we should try to move away from it.  This comment was a helpful reminder to create an issue for that: [DM-9592](https://jira.lsstcorp.org/browse/DM-9592).

File: python/lsst/meas/modelfit/optimizer/optimizer.cc [hunk 30]
```diff
@@ -0,0 +1,207 @@
...
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionInitialSize);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowReductionRatio);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowStepFraction);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowFactor);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionShrinkReductionRatio);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionShrinkFactor);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionSolverTolerance);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, maxInnerIterations);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, maxOuterIterations);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, doSaveIterations);
```
Comment (TallJimbo) [inline python/lsst/meas/modelfit/optimizer/optimizer.cc]: Happily most of these will probably never be changed - I just figured it was better to put them in the config class than define them as class- or module-level constants, just in case.  And the docs aren't great, but they do exist - they're just in the strings in the macros rather than comments.

File: python/lsst/meas/modelfit/truncatedGaussian.cc (hunk 23 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/truncatedGaussian.cc]: Yeah, I thought we had order rules that at most involved standard headers vs. third party vs. our own. Ugh, another place where our standards our both out of date and ignored, and I don't think the team (let alone me) has the energy to fix all of those.

File: python/lsst/meas/modelfit/optimizer/optimizer.cc (hunk 30 above)
Comment (r-owen) [inline python/lsst/meas/modelfit/optimizer/optimizer.cc]: My mistake. I missed the doc strings the first time (maybe I was looking at the defaults; I sure wish default values lived with the rest of the definition in the call to `LSST_CONTROL_FIELD`)

Comment (TallJimbo) [inline python/lsst/meas/modelfit/optimizer/optimizer.cc]: Ooh, that wasn't possible back when that macro was written, but I think we could do that with C++11.  I've added that to [DM-9222](https://jira.lsstcorp.org/browse/DM-9222).

File: python/lsst/meas/modelfit/model.cc (hunk 1 above)
Comment (pschella) [inline python/lsst/meas/modelfit/model.cc]: Yes indeed. My reading of the new rules was however that functions returning by (const) ref would be mandated to have a return value policy specified. But again, I think that is inconsistently applied anyway and may not be practical in such cases (with stl containers that are copied regardless). Your choice.

File: python/lsst/meas/modelfit/mixture.cc (hunk 4 above)
Comment (pschella) [inline python/lsst/meas/modelfit/mixture.cc]: Personally I would be happy with the standard that `__str__` returns whatever `<<` returns and `__repr__` just prints the name of the type. I'll put that on the CLO discussion and perhaps start an RFC (unless people push back on that suggestion).
I think this convention will make the experience of working with the code much nicer for users.
I don't think it is possible, or desirable to make `__repr__` work with `eval`.

File: tests/testProjectedLikelihood.py (hunk 9 above)
Comment (pschella) [inline tests/testProjectedLikelihood.py]: Fair enough. I discussed this with @jdswinbank yesterday. We should do a performance regression test at some point (probably after merge) and identify potential pain points.
If it turns out we need something in between that looks more like exposing `vector` as a view then we can always do so. But if we can get around doing that I think that would be a win.

File: python/lsst/meas/modelfit/model.cc (hunk 1 above)
Comment (TallJimbo) [inline python/lsst/meas/modelfit/model.cc]: I've added the return value policies.

File: python/lsst/meas/modelfit/mixture.cc (hunk 4 above)
Comment (r-owen) [inline python/lsst/meas/modelfit/mixture.cc]: I agree that in general it is not practical to make `__repr__` work with `eval`, but for simple cases when that works, then that is a useful goal. I like the idea of having `__str__` match `ostream <<` as the usual rule.
